### PR TITLE
Fix #154 - Add base class for transformers

### LINKED ISSUE #154: Parent class for transformers
@koaning Don't you think it's the time for a bit of refactoring and adding a parent `Transformer` class for all the transformers to subclass from it? For example, all of them have the same `__call__` method implementation. And also I think this makes the code more structured as it's getting bigger.

Willing to work on it if you are agreed.

### PR DESCRIPTION
**Pull request by mkaze**
This PR fixes #154 by adding an abstract base class for transformers and also performing some refactoring (i.e. removing `__call__` method implementation from subclasses).

### CODE REVIEW

File: whatlies/transformers/pca.py [hunk 1]
```diff
@@ -39,11 +40,6 @@ def __init__(self, n_components=2, **kwargs):
         self.kwargs = kwargs
```
Comment (koaning) [inline whatlies/transformers/pca.py:40]: Can't we remove the `self.is_fitted` here now?

File: whatlies/transformers/addrandom.py [hunk 2]
```diff
@@ -35,11 +36,6 @@ def __init__(self, n=1, sigma=0.1, seed=42):
         self.seed = seed
         self.is_fitted = False
```
Comment (koaning) [inline whatlies/transformers/addrandom.py:38]: I think we can remove this `self.is_fitted` now.

File: whatlies/transformers/ivis.py [hunk 3]
```diff
@@ -48,11 +49,6 @@ def __init__(self, n_components=2, **kwargs):
         self.kwargs["verbose"] = 0
         self.tfm = IVIS(embedding_dims=self.n_components, **self.kwargs)
```
Comment (koaning) [inline whatlies/transformers/ivis.py:50]: Can we remove the `self.is_fitted` now in `__init__`?

File: whatlies/transformers/noise.py [hunk 4]
```diff
@@ -38,11 +39,6 @@ def __init__(self, sigma=0.1, seed=42):
             lambda X: X + np.random.normal(0, sigma, X.shape)
```
Comment (koaning) [inline whatlies/transformers/noise.py:39]: Can we remove the `self.is_fitted` now in `__init__`?

File: whatlies/transformers/opentsne.py [hunk 5]
```diff
@@ -51,12 +52,6 @@ def __init__(self, n_components=2, **kwargs):
         self.kwargs = kwargs
```
Comment (koaning) [inline whatlies/transformers/opentsne.py:52]: Can we remove the `self.is_fitted` now in `__init__`?

File: whatlies/transformers/tsne.py [hunk 6]
```diff
@@ -43,11 +44,6 @@ def __init__(self, n_components=2, **kwargs):
         self.kwargs = kwargs
```
Comment (koaning) [inline whatlies/transformers/tsne.py:44]: Can we remove the `self.is_fitted` now in `__init__`?

File: whatlies/transformers/umap.py [hunk 7]
```diff
@@ -41,11 +42,6 @@ def __init__(self, n_components=2, **kwargs):
         self.kwargs = kwargs
```
Comment (koaning) [inline whatlies/transformers/umap.py:42]: Can we remove the `self.is_fitted` now in `__init__`?

**Review by koaning — CHANGES_REQUESTED** [review]
There might be some more code that we can remove now. Other than that -> this is sweet. 

In the future we can probably even extend this by making a parent class for all the scikit-learn based transformers but this is a great first cleanup!

**Comment by mkaze** [comment]
Actually, to remove the `self.is_fitted = False` in `__init__` method of subclasses they should call the base class constructure instead. I didn't do this and instead just added `self.is_fitted = False` to `__init__` method to make the (future!) developers aware that there is this property which should be maintained! I didn't like it even then, though! :smile: 

In other words, to initialize `self.is_fitted` either:

- `self.is_fitted = False` should be present in `__init__` method of all subclasses, or
- `super().__init__()` should be present in `__init__` method of all subclasses.

Of course, the second option is logical to me. @koaning Now, what do you think?

**Comment by koaning** [comment]
@mkaze If we're going to be refactoring I'd say the `super().__init__()` is going to be the way forward.

**Review by koaning — APPROVED** [review]

**Comment by koaning** [comment]
LGTM

**Pull request merged**